src/GridFS/GridFsUpload.php

```diff
@@ -0,0 +1,169 @@
+    private $length=0;
+    /**
+     * Constructs a GridFS upload stream
+     *
+     * Supported options:
+     *
+     *  * contentType (string): DEPRECATED content type to be stored with the file.
+     *    This information should now be added to the metadata
+     *
+     *  * aliases (array of strings): DEPRECATED An array of aliases.
+     *    Applications wishing to store aliases should add an aliases field to the
+     *    metadata document instead.
+     *
+     *  * metadata (array or object): User data for the 'metadata' field of the files
+     *    collection document.
+     *
+     *  * writeConcern (MongoDB\Driver\WriteConcern): Write concern.
+     *
+     * @param \MongoDB\Collection    $filesCollection  Files Collection
+     * @param \MongoDB\Collection    $chunksCollection Chunks Collection
+     * @param int32                  $chunkSizeBytes   Size of chunk
+     * @param string                 $filename         Filename to insert
+     * @param array                  $options          File options
+     * @throws InvalidArgumentException
+     */
+    public function __construct(
+        Bucket $bucket,
+        $filename,
+        array $options=[]
+        )
+    {
+        $this->bufferLength = 0;
+        $this->ctx = hash_init('md5');
+
+        $uploadDate = time();
+        $objectId = new \MongoDB\BSON\ObjectId();
+        $main_file = [
+            "chunkSize" => $bucket->getChunkSizeBytes(),
+            "filename" => $filename,
+            "uploadDate" => $uploadDate,
+            "_id" => $objectId
+        ];
+
+        $fileOptions = [];
+        if (isset($options['contentType'])) {
+            if (is_string($options['contentType'])) {
+                $fileOptions['contentType'] = $options['contentType'];
+            } else {
+                throw new InvalidArgumentTypeException('"contentType" option', $options['contentType'], 'string');
+            }
+        }
+        if (isset($options['aliases'])) {
+            if (\MongoDB\is_string_array($options['aliases'])) {
+                $fileOptions['aliases'] = $options['aliases'];
+            } else {
+                throw new InvalidArgumentTypeException('"aliases" option', $options['aliases'], 'array of strings');
+            }
+        }
+
+        if (isset($options['metadata'])) {
+            if (is_array($options['metadata']) || is_object($options['metadata'])) {
+                $fileOptions['metadata'] = $options['metadata'];
+            } else {
+                throw new InvalidArgumentTypeException('"metadata" option', $options['metadata'], 'object or array');
+            }
+        }
+        $this->file = array_merge($main_file, $fileOptions);
+        parent::__construct($bucket);
+    }
+    /**
+    * Reads data from a stream into GridFS
+    *
+    * @param Stream   $source   Source Stream
+    * @return ObjectId
+    */
+    public function uploadFromStream($source)
+    {
+        $this->bucket->ensureIndexes();
```

Comment:
It would appear to me, upon re-examining the [documentation](https://github.com/mongodb/specifications/blob/master/source/gridfs/gridfs-spec.rst#before-write-operations) that it cannot.